llama-index-core/llama_index/core/base/base_retriever.py

```diff
@@ -166,12 +166,12 @@ def _handle_recursive_retrieval(
             else:
                 retrieved_nodes.append(n)
 
-        seen = set()
-        return [
-            n
-            for n in retrieved_nodes
-            if not (n.node.hash in seen or seen.add(n.node.hash))  # type: ignore[func-returns-value]
-        ]
+        res = dict()
+        for n in retrieved_nodes:
+            node_hash = n.node.hash
+            if node_hash not in res:
+                res[node_hash] = n
+        return list(res.values()]
```

Comment:
can you provide some motivation for this change? Using a `set()` like before should be just fine

Was this tested? I see a stray bracket at the end of the line here too

Reply:
I think using a dictionary is more readable and simpler than using a set

Reply:
> can you provide some motivation for this change? Using a `set()` like before should be just fine
> 
> Was this tested? I see a stray bracket at the end of the line here too

You are right, there is indeed an extra bracket at the end

Reply:
.... can you remove it? :)